### PR TITLE
Requirements Parsing and Unit Tests

### LINKED ISSUE #25: OSM SkillEntry.requirements is list not dict
When parsing a skill with requirements.txt, `SkillEntry._data["requirements"] = {list_parsed_from_requirements_txt}`. At install, `self.requirements` is expected to be a dict of "python", "skill", "system" dependencies.

Few considerations here:
* It appears that `requirements.txt` is overriding the `skill.json`. In general, should repo files take priority over `skill.json`?

* `requirements.txt` is superseded by `requirements.yml`, should we handle cases where both of these exist? How do these interact with `skill.json`?

### PR DESCRIPTION
**Pull request by NeonDaniel**
Updates requirements parsing to merge requirements from `skill.json`, `manifest.yml`, and `requirements.txt` in priority order
Updates existing unit tests and adds more `SkillEntry` tests

Closes #25 #24 

### TODO:
- [x] ~~Test and handle conflicting dependency versions~~
- [x] Test commented lines in requirements
- [x] Format git spec'd dependencies with auth_key (Maybe this is installer scope and not SkillEntry)
- [x] Update existing references to test case where possible

### CODE REVIEW

File: ovos_skills_manager/licenses.py [hunk 1]
```diff
@@ -140,6 +140,7 @@ def _is_mit(lic):
 
 
 def parse_license_type(lic):
+    # TODO: If this is a github repo, they have an api: https://docs.github.com/en/rest/reference/licenses
```
Comment (JarbasAl) [inline ovos_skills_manager/licenses.py:143]: there are methods in the github module for this, this is a separate utility to parse text

Eg, a local skill folder

File: tests/test_github_branch_parsing.py [hunk 2]
```diff
@@ -14,7 +16,7 @@
 
 class TestGithubBranchParsing(unittest.TestCase):
 
-    def test_url(self):
+    def test_branch_from_url(self):
         url_no_branch = "https://github.com/JarbasSkills/skill-ddg"
         url_branch = "https://github.com/JarbasSkills/skill-ddg/tree/v0.1.0"
```
Comment (JarbasAl) [inline tests/test_github_branch_parsing.py:21]: these should also use a test repo setup explicitly for this purpose

**Comment by JarbasAl** [comment]
i assume #10 will be incorporated in this PR ?

File: ovos_skills_manager/licenses.py (hunk 1 above)
Comment (NeonDaniel) [reply ovos_skills_manager/licenses.py:143]: Okay, I missed the github methods. I'll remove this TODO

**Comment by NeonDaniel** [comment]
> i assume #10 will be incorporated in this PR ?

Should that be merged first so this one just adds it to automation, or should I just integrate those changes here?

**Comment by JarbasAl** [comment]
> > i assume #10 will be incorporated in this PR ?
> 
> Should that be merged first so this one just adds it to automation, or should I just integrate those changes here?

feels cleaner to do it in a single PR i think, but feel free to push the merge button on #10 if it makes your life easier!

**Comment by NeonDaniel** [comment]
> > > i assume #10 will be incorporated in this PR ?
> > 
> > 
> > Should that be merged first so this one just adds it to automation, or should I just integrate those changes here?
> 
> feels cleaner to do it in a single PR i think, but feel free to push the merge button on #10 if it makes your life easier!

I'll pull those commits in here then, easier to test with automation that way

File: ovos_skills_manager/osm.py [hunk 3]
```diff
@@ -210,7 +210,7 @@ def search_skills_by_url(self, url, as_json=False):
             store.authenticate()
             for skill in store.search_skills_by_url(url, as_json):
                 store.clear_authentication()
-                return skill
+                return [skill]
```
Comment (NeonDaniel) [inline ovos_skills_manager/osm.py:213]: This matches behavior of the other search methods, but could break references (like the one in NeonCore skill_store). I think it is better to match the other search methods and avoid changes in the search script, but could see arguments either way.

Should this change be reverted and function calls updated to handle this return as a string?

File: tests/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import os
+import sys
+import unittest
+
+sys.path.append(os.path.dirname(os.path.dirname(__file__)))
```
Comment (JarbasAl) [inline tests/test_utils.py:5]: why is this needed?

File: ovos_skills_manager/osm.py (hunk 3 above)
Comment (JarbasAl) [reply ovos_skills_manager/osm.py:213]: i think all search methods should be consistent, this seems like an oversight and not a design decision, if there was a reason to return a string i cant remember it... so i'd say this is a bug fix

File: tests/test_utils.py (hunk 4 above)
Comment (NeonDaniel) [reply tests/test_utils.py:5]: For github automation, the package isn't actually installed, so this is so we can import from the cloned repository

File: ovos_skills_manager/osm.py (hunk 3 above)
Comment (NeonDaniel) [reply ovos_skills_manager/osm.py:213]: I'm guessing it was a speedup for when we have a literal URL, but I agree.

File: ovos_skills_manager/appstores/__init__.py [hunk 5]
```diff
@@ -69,7 +72,8 @@ def sync_skills_list(self, merge=False, new_only=False):
         if new_only:
             skiplist = [s["url"] for s in self.db if s.get("url")]
 
-        for skill in self.get_skills_list(skiplist=skiplist):
+        skills = self.get_skills_list(skiplist=skiplist) or []
+        for skill in skills:
```
Comment (ChanceNCounter) [inline ovos_skills_manager/appstores/__init__.py:76]: FWIW, this statement can still be fed to `for` without the intermediate assignment. Excuse IPython:

```python
In [4]: [item for item in list(range(0)) or [1, 2, 3]]
Out[4]: [1, 2, 3]

**Comment by ChanceNCounter** [comment]
Closes #24

File: ovos_skills_manager/appstores/mycroft_marketplace.py [hunk 6]
```diff
@@ -10,8 +10,11 @@
...
 def get_current_marketplace_branch():
     # TODO check mycroft version for default branch as fallback
-    default_branch = read_mycroft_config().get("skills", {}) \
-        .get("msm", {}).get("repo", {}).get("branch", "20.08")
+    try:
+        default_branch = read_mycroft_config().get("skills", {}) \
+            .get("msm", {}).get("repo", {}).get("branch", "20.08")
+    except FileNotFoundError:
+        default_branch = "20.08"
     return default_branch
```
Comment (ChanceNCounter) [inline ovos_skills_manager/appstores/mycroft_marketplace.py:18]: Technically OOS, but this is probably a good time to bump that number.

Also, at some point, we should try to resolve this paradox and stop hardcoding this number. It's hard to keep track of things like this.

Comment (ChanceNCounter) [reply ovos_skills_manager/appstores/mycroft_marketplace.py:18]: And just from the Pythonic perspective (I don't really care =P)

```python
try:
    return read_mycroft_config()...
except FileNotFoundError:
    return "21.02"
```

File: ovos_skills_manager/github/__init__.py [hunk 7]
```diff
@@ -110,6 +109,7 @@ def get_skill_data(url, branch=None):
     # augment with json data
     # this should take precedence over everything else
     try:
+        data["requirements"]["system"] = {k: v.split() for k, v in data["requirements"].get("system", {}).items()}
```
Comment (ChanceNCounter) [inline ovos_skills_manager/github/__init__.py:112]: I'm pretty sure I know what's going on here, but I'm not sure how it could've been masked earlier (unless it was *never* taking precedence) so I'm asking anyway =P

File: ovos_skills_manager/github/raw.py [hunk 8]
```diff
@@ -457,6 +457,8 @@ def get_android_json_from_github_url(url, branch=None):
 def get_branch_from_skill_json_github_url(url, branch=None):
     try:
         branch = branch or get_branch_from_github_url(url)
+        if '@' in url:
+            url = url.split('@', 1)[0]
```
Comment (ChanceNCounter) [inline ovos_skills_manager/github/raw.py:461]: I feel like there's an edge case here, but it isn't coming to me.

File: ovos_skills_manager/skill_entry.py [hunk 9]
```diff
@@ -320,4 +320,9 @@ def is_previously_installed(self, folder=None):
         return isdir(join(folder, self.uuid))
 
     def __repr__(self):
+        if not self.skill_name:
+            return self.url
         return self.skill_name + " " + self.url
+
+    def __eq__(self, other):
+        return self.json == other.json
```
Comment (ChanceNCounter) [inline ovos_skills_manager/skill_entry.py:328]: thank you please!

File: ovos_skills_manager/utils.py [hunk 10]
```diff
@@ -1,3 +1,19 @@
...
+    """
+    # Handle case sensitivity in dependencies and any potentially required auth
+    for i in range(0, len(python_deps)):
+        r = python_deps[i]
+        if "@" in r:
+            parts = [p.lower() if p.strip().startswith("git+http") else p for p in r.split('@')]
+            r = "@".join(parts)
+        if token:
+            if "github.com" in r:
+                r = r.replace("github.com", f"{token}@github.com")
```
Comment (ChanceNCounter) [inline ovos_skills_manager/utils.py:13]: These warrant descriptive comments for posterity

File: ovos_skills_manager/appstores/__init__.py (hunk 5 above)
Comment (NeonDaniel) [reply ovos_skills_manager/appstores/__init__.py:76]: I tend to prefer defining outside of loop statements for easier debugging, but I can make this change if concise code is preferred here.

File: ovos_skills_manager/appstores/mycroft_marketplace.py (hunk 6 above)
Comment (NeonDaniel) [reply ovos_skills_manager/appstores/mycroft_marketplace.py:18]: Ideally, I would like to see a static reference to the latest too, though I think that is up to Mycroft unless we want to maintain a fork that points at 'latest'.

I also agree with the second comment; more correct but immaterial IMO

File: ovos_skills_manager/github/__init__.py (hunk 7 above)
Comment (NeonDaniel) [reply ovos_skills_manager/github/__init__.py:112]: requirements per the Mycroft `manifest.yml` is a string that is passed directly to `apt`/whatever package manager so far as I can tell.. I split it here to merge it with `skill.json` the same way python deps are handled

File: ovos_skills_manager/github/raw.py (hunk 8 above)
Comment (NeonDaniel) [reply ovos_skills_manager/github/raw.py:461]: I suppose if the url includes auth this could break, though that is depreciated behavior now.. Other than that `@` isn't a valid character in any github repo address (almost positive).

File: ovos_skills_manager/utils.py (hunk 10 above)
Comment (NeonDaniel) [reply ovos_skills_manager/utils.py:16]: Annotated, documented, and commented.

File: ovos_skills_manager/appstores/__init__.py (hunk 5 above)
Comment (ChanceNCounter) [reply ovos_skills_manager/appstores/__init__.py:76]: Nope, just checking.

File: ovos_skills_manager/appstores/mycroft_marketplace.py (hunk 6 above)
Comment (ChanceNCounter) [reply ovos_skills_manager/appstores/mycroft_marketplace.py:18]: Well, we should at least add TODOs, or pin an issue, or do something to make sure we keep this number up to date (already outdated)

**Review by ChanceNCounter — APPROVED** [review]
I'm still concerned about the relative sluggishness of the github parser, but that's out of scope.

I do think the hardcoded Mycroft-core version (see currently-unresolved box above) should be bumped here, but otherwise, the little checklist in my PR extension says this is done :)

**Comment by NeonDaniel** [comment]
Should I increment version in this PR?
